### DISCUSSION TITLE
Any way to collect / collate results of multiple machines in an Array?

### DISCUSSION THREAD

**Discussion by ratacat** [statelyai/xstate#1853] [Q&A]
I have something like this where I'm running hundreds of these machines in a batch at once, and I'd love to be able to access the event.data via promise or something so I could wait for all of them to finish.

```
let buyAmountMachine = createBuyAmountMachineWithContext(
          { Pair, TokenAmount, Route, weth, tx, logger },
          tx.hash.slice(-5),
        )
        const service = interpret(buyAmountMachine)
        buyAmountMachinesList.set(service.id, service)
        service.onDone((event) => {
          //logger.info(`${service.id} finished!`)
          buyAmountMachinesList.delete(service.id)
          //the data I want is from the finishing event, is here
          event.data.calc
        }
```

Any thoughts? For the moment, I've basically resorted to pushing to an array located one level up in the scope, and then looping through it until they are all received.

**Comment by davidkpiano** [comment]
> Any thoughts? For the moment, I've basically resorted to pushing to an array located one level up in the scope, and then looping through it until they are all received.

This makes sense, IMO. It models how you would do it in real-life, as well.

You can have a top-level transition like:

```js
always: {
  target: 'allDone',
  cond: 'allThingsDone'
}
```

Where `'allThingsDone'` is implemented like `ctx => ctx.things.every(...)`.

**Reply by ratacat** [reply]
Ah, so in this instance, my top level control flow is actually not in xstate. But it is managing a number of different xstate machines.
